File: utils/markup.py
import re


def strip_markup(text: str) -> str:
    """Strip the 5eTools-style inline markup from *text*.

    Handles patterns such as:

    - "{@variantrule Hit Points|XPHB}" -> "Hit Points"
    - "{@variantrule Emanation [Area of Effect]|XPHB|Emanation}" -> "Emanation"
    - "{@damage 1d6}" -> "1d6"

    If ``text`` is falsy ("" or ``None``), it is returned unchanged.
    """

    if not text:
        return text

    # First handle commands which include a trailing "|source" (and possibly
    # more pipe‑separated segments). We keep only the main display text before
    # any optional bracketed qualifier like " [Area of Effect]".
    #
    # Regex breakdown:
    #   \{@          - opening marker
    #   ([^\s}]+)    - command name (not used)
    #   \s+          - at least one space
    #   ([^|}\[]+)   - main display text up to '|' or ' ['
    #   (?:\s*\[[^]]+])? - optional " [kind]" section to be discarded
    #   \|          - separator before source/extra segments
    #   [^}]*        - the rest up to
    #   }            - closing brace
    pattern_with_source = re.compile(
        r"\{@([^\s}]+)\s+([^|}\[]+)(?:\s*\[[^]]+])?\|[^}]*}"
    )

    def _repl_with_source(match: re.Match) -> str:
        return match.group(2).strip()

    cleaned = pattern_with_source.sub(_repl_with_source, text)

    # Then handle simpler commands with no "|source" part, such as
    # "{@damage 1d6}". In this case we keep everything after the command
    # name up to the closing brace.
    pattern_simple = re.compile(r"\{@([^\s}]+)\s+([^}]+)}")

    def _repl_simple(match: re.Match) -> str:
        return match.group(2).strip()

    return pattern_simple.sub(_repl_simple, cleaned)

File: src/markup.rs
use regex::Regex;

pub fn strip_markup(text: Option<&str>) -> Option<String> {
    let Some(text) = text else { return None };
    if text.is_empty() {
        return Some(text.to_string());
    }

    // {@scaledamage 2d8|1-9|1d8} -> 1d8
    let pattern_scaledamage =
        Regex::new(r"\{@scaledamage\s+[^|}]+\|[^|}]*\|([^}]+)}").expect("valid regex");

    // {@variantrule Hit Points|XPHB} -> Hit Points
    // {@variantrule Emanation [Area of Effect]|XPHB|Emanation} -> Emanation
    let pattern_with_source =
        Regex::new(r"\{@([^\s}]+)\s+([^|}\[]+)(?:\s*\[[^]]+])?\|[^}]*}").expect("valid regex");

    // {@damage 1d6} -> 1d6
    let pattern_simple = Regex::new(r"\{@([^\s}]+)\s+([^}]+)}").expect("valid regex");

    let cleaned = pattern_scaledamage
        .replace_all(text, |caps: &regex::Captures<'_>| {
            caps[1].trim().to_string()
        })
        .to_string();

    let cleaned = pattern_with_source
        .replace_all(&cleaned, |caps: &regex::Captures<'_>| {
            caps[2].trim().to_string()
        })
        .to_string();

    let cleaned = pattern_simple
        .replace_all(&cleaned, |caps: &regex::Captures<'_>| {
            caps[2].trim().to_string()
        })
        .to_string();

    Some(cleaned)
}

pub fn strip_markup_or_empty(text: &str) -> String {
    strip_markup(Some(text)).unwrap_or_default()
}
